test/extensions/transport_sockets/alts/tsi_socket_test.cc

```diff
@@ -628,6 +650,252 @@ TEST_F(TsiSocketTest, DoReadDrainBufferTwice) {
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer, bool) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, buffer.length(), false};
+        client_to_server_.move(buffer);
+        return result;
+      }));
+  expectIoResult({Network::PostIoAction::KeepOpen, 17UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+  EXPECT_EQ(makeFakeTsiFrame(ClientToServerDataFirstHalf) +
+                makeFakeTsiFrame(ClientToServerDataSecondHalf),
+            client_to_server_.toString());
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_)).WillOnce(Invoke([&](Buffer::Instance& buffer) {
+    Network::IoResult result = {Network::PostIoAction::Close, client_to_server_.length(), false};
+    buffer.move(client_to_server_);
+    return result;
+  }));
+  expectIoResult({Network::PostIoAction::Close, 17UL, false},
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+  EXPECT_EQ(ClientToServerData, server_.read_buffer_.toString());
+}
+
+TEST_F(TsiSocketTest, DoWriteOutstandingHandshakeData) {
+  auto validator = [](const tsi_peer&, std::string&) { return true; };
+  initialize(validator, validator);
+
+  InSequence s;
+  doFakeInitHandshake();
+
+  EXPECT_CALL(*client_.raw_socket_, doRead(_));
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 0UL, false},
+                 client_.tsi_socket_->doRead(client_.read_buffer_));
+  EXPECT_EQ(makeFakeTsiFrame("CLIENT_FINISHED"), client_to_server_.toString());
+  EXPECT_EQ(0L, client_.read_buffer_.length());
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_));
+  EXPECT_CALL(server_.callbacks_, raiseEvent(Network::ConnectionEvent::Connected));
+
+  // Write the first part of handshake data (14 bytes).
+  EXPECT_CALL(*server_.raw_socket_, doWrite(_, false))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer, bool) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, buffer.length() - 5, false};
+        server_to_client_.add(buffer.linearize(0), buffer.length() - 5);
+        buffer.drain(buffer.length() - 5);
+        return result;
+      }));
+  EXPECT_CALL(*server_.raw_socket_, doRead(_));
+  expectIoResult({Network::PostIoAction::KeepOpen, 0UL, false},
+                 server_.tsi_socket_->doRead(server_.read_buffer_));
+
+  EXPECT_EQ(makeFakeTsiFrame("SERVER_FINISHED").length(), 19);
+  EXPECT_EQ(makeFakeTsiFrame("SERVER_FINISHED").substr(0, 14), server_to_client_.toString());
+
+  server_.write_buffer_.add(ServerToClientData);
+  server_.tsi_socket_->setActualFrameSizeToUse(LargeFrameSize);
+
+  // Write the second part of handshake data (4 bytes).
+  EXPECT_CALL(*server_.raw_socket_, doWrite(_, false))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer, bool) {
+        Network::IoResult result = {Network::PostIoAction::KeepOpen, 4, false};
+        server_to_client_.add(buffer.linearize(0), 4);
```

Comment:
I don't think 0 is a valid argument to linearize.  I think you need linearize(4) to match the number of bytes you're adding.

Alternatively, consider:
`server_to_client_.move(buffer, 4);`

which is a nicer version of add(..., 4) followed by drain(4)

Reply:
Nice suggestion. The code is much simpler now.